core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider_life_cycle.go

```diff
@@ -17,42 +18,113 @@ var (
-		return errors.Errorf("filter for upkeep with id %s already registered", upkeepID.String())
+	var filter upkeepFilter
+	currentFilter := p.filterStore.Get(upkeepID)
+	if currentFilter != nil {
+		if currentFilter.configUpdateBlock > opts.UpdateBlock {
+			// already registered with a config from a higher block number
+			return fmt.Errorf("filter for upkeep with id %s already registered with newer config", upkeepID.String())
+		} else if currentFilter.configUpdateBlock == opts.UpdateBlock {
+			// already registered with the same config
+			p.lggr.Debugf("filter for upkeep with id %s already registered with the same config", upkeepID.String())
+			return nil
+		}
+		// removing filter so we can recreate it with updated values
+		err := p.poller.UnregisterFilter(p.filterName(currentFilter.upkeepID))
+		if err != nil {
+			return fmt.Errorf("failed to unregister upkeep filter %s for update: %w", upkeepID.String(), err)
+		}
+		filter = *currentFilter
+	} else { // new filter
+		filter = upkeepFilter{
+			upkeepID:     upkeepID,
+			blockLimiter: rate.NewLimiter(p.opts.BlockRateLimit, p.opts.BlockLimitBurst),
+		}
 	}
-	// TODO: ReplayAsync logs for the upkeep from creation block
-	// TODO: Verify handling of different upkeeps using same filter
-	if err := p.poller.RegisterFilter(filter); err != nil {
-		return errors.Wrap(err, "failed to register upkeep filter")
+	filter.lastPollBlock = 0
+	filter.lastRePollBlock = 0
+	filter.configUpdateBlock = opts.UpdateBlock
+	filter.addr = lpFilter.Addresses[0].Bytes()
+	filter.topics = make([]common.Hash, len(lpFilter.EventSigs))
+	copy(filter.topics, lpFilter.EventSigs)
+
+	if err := p.register(lpFilter, filter); err != nil {
+		return fmt.Errorf("failed to register upkeep filter %s: %w", filter.upkeepID.String(), err)
 	}
-	topics := make([]common.Hash, len(filter.EventSigs))
-	copy(topics, filter.EventSigs)
-	p.filterStore.AddActiveUpkeeps(upkeepFilter{
-		upkeepID:      upkeepID,
-		addr:          filter.Addresses[0].Bytes(),
-		topics:        topics,
-		blockLimiter:  rate.NewLimiter(p.opts.BlockRateLimit, p.opts.BlockLimitBurst),
-		lastPollBlock: 0, // TODO: Start from filter creation block
-	})
+
+	return nil
+}
+
+// register registers the upkeep filter with the log poller and adds it to the filter store.
+func (p *logEventProvider) register(lpFilter logpoller.Filter, ufilter upkeepFilter) error {
+	if err := p.poller.RegisterFilter(lpFilter); err != nil {
+		return err
+	}
+	p.filterStore.AddActiveUpkeeps(ufilter)
+	p.poller.ReplayAsync(int64(ufilter.configUpdateBlock))
 
 	return nil
 }
 
 func (p *logEventProvider) UnregisterFilter(upkeepID *big.Int) error {
 	err := p.poller.UnregisterFilter(p.filterName(upkeepID))
-	if err == nil {
-		p.filterStore.RemoveActiveUpkeeps(upkeepFilter{
-			upkeepID: upkeepID,
-		})
+	if err != nil {
+		// TODO: mark as removed in filter store, so we'll
+		// automatically retry on next refresh
```

Comment:
this would automatically be handled now by our RefreshActiveUpkeeps logic upon next refresh?